File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int calculateDifference(vector<int>& arr) {
    int sum = 0;
    for (int i = 0; i < arr.size() - 1; ++i) {
        sum += abs(arr[i] - arr[i+1]);
    }
    return sum;
}

int main() {
    int N;
    cin >> N;
    vector<int> A(N);
    for (int i = 0; i < N; ++i) {
        cin >> A[i];
    }

    // 모든 순열에 대해 계산
    sort(A.begin(), A.end()); // 순열을 만들기 위해 정렬
    int maxDifference = 0;
    do {
        // 현재 순열에 대한 차이값 합 계산
        int currentDifference = calculateDifference(A);
        maxDifference = max(maxDifference, currentDifference);
    } while (next_permutation(A.begin(), A.end())); // 다음 순열 생성

    cout << maxDifference << endl;
    return 0;
}

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {

    private static int N, answer;
    private static int[] arr, result;
    private static boolean[] visited;


    private static int calMax(int[] result) {
        int sum = 0;
        for (int i = 0; i < N-1; i++) {
            sum += Math.abs(result[i] - result[i+1]);
        }
        return sum;
    }
    private static void dfs(int depth) {
        //2-1. 종료조건: N개를 다 뽑은 경우
        if (depth == N) {
            //3. 최대값인지 확인 후 갱신하기
            answer = Math.max(answer, calMax(result));
            return;
        }
        //2-2. 순열 & 백트래킹
        for (int i = 0; i < N; i++) {
            if (!visited[i]) {
                visited[i] = true;
                result[depth] = arr[i];
                dfs(depth + 1);
                visited[i] = false;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        //1. 입력받기
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        arr = new int[N];
        result = new int[N];
        visited = new boolean[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        //2. N개의 순열을 재귀로 호출
        dfs(0);

        System.out.println(answer);
    }
}